src/main/java/com/labsynch/labseer/service/LsThingServiceImpl.java

```diff
@@ -183,74 +149,106 @@ public PreferredNameResultsDTO getCodeNameFromName(String thingType, String thin
+		predicates = predicateList.toArray(predicates);
+		// Construct the query to fetch back LsThings
+		criteria.select(lsThingRoot);
+		criteria.distinct(true);
+		// AND all the terms together
+		criteria.where(cb.and(predicates));
+		TypedQuery<LsThing> q = em.createQuery(criteria);
+		logger.debug(q.unwrap(org.hibernate.Query.class).getQueryString());
+		List<LsThing> foundLsThings = q.getResultList();
+		// Construct a PreferredNameDTO for each possible requestName for each found LsThing, both by label and by codeName
+		// The goal is to be able to line up with what was sent in, whether it was by label or by codeName
+		List<PreferredNameDTO> rawResultDTOs = new ArrayList<PreferredNameDTO>();
+		for (LsThing lsThing : foundLsThings){
+			String bestLabel = pickBestLabel(lsThing);
+			String codeName = lsThing.getCodeName();
+			for (LsThingLabel label : lsThing.getLsLabels()){
+				boolean typeMatches = labelType == null || label.getLsType().equals(labelType);
+				boolean kindMatches = labelKind == null || label.getLsKind().equals(labelKind);
+				if (typeMatches && kindMatches){
+					PreferredNameDTO resByLabel = new PreferredNameDTO(label.getLabelText(), bestLabel, codeName);
+					rawResultDTOs.add(resByLabel);
 				}
 			}
+			PreferredNameDTO resByCodeName = new PreferredNameDTO(codeName, bestLabel, codeName);
+			rawResultDTOs.add(resByCodeName);
 		}
-		responseOutput.setResults(requests);
-		responseOutput.setErrorMessages(errors);
-
-		return responseOutput;
-	}
-	
-	
-	public PreferredNameResultsDTO getCodeNameFromName(PreferredNameRequestDTO requestDTO){
-
-		logger.info("number of requests: " + requestDTO.getRequests().size());
-		Collection<PreferredNameDTO> requests = requestDTO.getRequests();
-
-		PreferredNameResultsDTO responseOutput = new PreferredNameResultsDTO();
-		Collection<ErrorMessageDTO> errors = new HashSet<ErrorMessageDTO>();
+		
 
+		// Work up query results into output
 		for (PreferredNameDTO request : requests){
-			request.setPreferredName("");
-			request.setReferenceName("");
-			List<LsThing> lsThings = LsThing.findLsThingByLabelText(request.getRequestName()).getResultList();
-			if (lsThings.size() == 1){
-				request.setPreferredName(lsThings.get(0).getCodeName());
-				request.setReferenceName(lsThings.get(0).getCodeName());
-			} else if (lsThings.size() > 1){
+			List<PreferredNameDTO> matches = new ArrayList<PreferredNameDTO>();
+			for (PreferredNameDTO res : rawResultDTOs){
+				if (res.getRequestName().equals(request.getRequestName())){
+					matches.add(res);
+				}
+			}
+			// Check for multiple or zero matches
+			if (matches.size() == 1){
+				PreferredNameDTO match = matches.get(0);
+				request.setPreferredName(match.getPreferredName());
+				request.setReferenceName(match.getReferenceName());
+			} else if (matches.size() > 1){
 				responseOutput.setError(true);
 				ErrorMessageDTO error = new ErrorMessageDTO();
-				error.setLevel("error");
+				error.setLevel("MULTIPLE RESULTS");
```

Comment:
I believe this should still be `error`

Reply:
Yeah this was inconsistent between two sections of code. I agree it should be `error`. I'll fix.